libs/frontend/util-command/src/lib/command.ts

```diff
@@ -8,6 +13,24 @@ import { undoCommand } from './undo-command'
 // Helper for pushing a command to separate siteStore.save
 const pushCommandHelper = (site: ISite, command: ICommand, clearRedo = true) => {
   const cmd = optimizeCommandGroup(command)
+  const { editingMixinData } = site.editor ?? {}
+
+  // Push a command to close mixin menu for user when necessary
+  if (editingMixinData && shouldCloseMixinMenuForUser(cmd)) {
+    const closeMixinMenuCommand: ICommand<IUpdateUiData<UiAction.CloseMixinMenu>> = {
+      type: CommandType.UpdateUi,
+      data: {
+        action: UiAction.CloseMixinMenu,
+        params: {
+          mixinId: editingMixinData?.mixinId as string,
+          originComponentId: editingMixinData?.originComponentId,
+        },
+      },
+    }
+    applyCommand(site, closeMixinMenuCommand)
+    site.history.back.push(closeMixinMenuCommand)
+  }
```

Comment:
The purpose of these code is to push a close-mixin-menu command to the history for user, if the mixin menu is currently opened, and the new command is not meant to close the menu. For example, if a user adds a new component to the editor by clicking Text in the submenu when the mixin menu is opened, the mixin menu should be closed, and the new component should be selected.

Instead of finding out all cases where the same scenario would happen and handle them one by one, I thought it would easier to just add it here, right before we push a new command. It feels hacky, but I didn't think of a better solution. 😬 

In addition, as you mentioned, there are still some cases that's not being handled. We can try to fix them in #304!